### PR TITLE
Implement indexing in ATen

### PR DESCRIPTION
**Pull request by colesbury**
```
Implements basic and advanced indexing using ATen tensors/variables.
Basic indexing is translated at the Python-binding level
(python_variable_indexing.cpp) to narrow/squeeze/unsqueeze/select calls.
Advanced indexing is implemented in ATen in terms of take() and put()
calls.
```

### CODE REVIEW

**Comment by colesbury** [comment]
FYI, `__setitem__` previously allowed the input and destination to be non-broadcastable, but have the same number of elements. This change makes `__setitem__` broadcast the input to the destination like we do with other functions.

File: tools/autograd/templates/python_variable_methods.cpp [hunk 1]
```diff
@@ -52,6 +52,24 @@ static PyObject * THPVariable_contiguous(PyObject* self, PyObject* args)
   END_HANDLE_TH_ERRORS
 }
 
+static PyObject * THPVariable_copy_(PyObject* self, PyObject* args, PyObject* kwargs)
+{
+  HANDLE_TH_ERRORS
+  static PythonArgParser parser({ "copy_(Tensor other)" });
+  auto& self_ = reinterpret_cast<THPVariable*>(self)->cdata;
```
Comment (killeent) [inline tools/autograd/templates/python_variable_methods.cpp]: Not sure what the semantics are here, but are we guaranteed that `self` is a THPVariable? i.e. why is there not a `THPVariable_Check(...)` here?

File: torch/csrc/autograd/python_variable_indexing.cpp [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+  }
+  // This uses a heuristics from NumPy for determining whether to treat
+  // non-tuple sequences as if they were a tuple. From the NumPy code comments:
+  //
+  // "At this point, we're left with a non-tuple, non-array, sequence:
+  //  typically, a list. We use some somewhat-arbitrary heuristics from here
+  //  onwards to decided whether to treat that list as a single index, or a
+  //  list of indices. Backwards compatibility only takes effect for short
+  //  sequences - otherwise we treat it like any other scalar."
+  auto n = PySequence_Size(index);
```
Comment (killeent) [inline torch/csrc/autograd/python_variable_indexing.cpp]: When would this be negative?

File: torch/csrc/autograd/python_variable_indexing.cpp [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+  if (!PySequence_Check(index)) {
+    return false;
+  }
+  // This uses a heuristics from NumPy for determining whether to treat
+  // non-tuple sequences as if they were a tuple. From the NumPy code comments:
+  //
+  // "At this point, we're left with a non-tuple, non-array, sequence:
+  //  typically, a list. We use some somewhat-arbitrary heuristics from here
+  //  onwards to decided whether to treat that list as a single index, or a
+  //  list of indices. Backwards compatibility only takes effect for short
```
Comment (killeent) [inline torch/csrc/autograd/python_variable_indexing.cpp]: What is the backwards compatibility they refer to?

File: torch/csrc/autograd/python_variable_indexing.cpp [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+  auto n = PySequence_Size(index);
+  if (n < 0) {
+    PyErr_Clear();
+    return false;
+  }
+  if (n >= 32) {
+    return false;
+  }
+  for (Py_ssize_t i = 0; i < n; i++) {
+    auto obj = THPObjectPtr{PySequence_GetItem(index, i)};
```
Comment (killeent) [inline torch/csrc/autograd/python_variable_indexing.cpp]: There was some debate when I did my original indexing diff on whether or not to use `FAST` sequencing indexing. I'll defer to @apaszke on this.

File: torch/csrc/autograd/python_variable_indexing.cpp [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+    if (obj.get() == Py_Ellipsis || obj.get() == Py_None) {
+      return true;
+    }
+  }
+  return false;
+}
+
+static THPObjectPtr wrapTuple(PyObject* index) {
+  THPObjectPtr res;
+  if (treatSequenceAsTuple(index)) {
```
Comment (killeent) [inline torch/csrc/autograd/python_variable_indexing.cpp]: I'm a little confused as to the distinction between when you would use `PySequence_Tuple` and `PyTuple_Pack`. Can you explain in more detail what is going on here? How come in the later case our tuple is of a single element?

File: torch/csrc/autograd/python_variable_indexing.cpp [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+      return wrap(self_.type().tensor({0}));
+    }
+  } else if (PySlice_Check(index)) {
+    return wrap(applySlice(self_, 0, index, true));
+  }
+
+  // wrap index in a tuple if it's not already one
+  THPObjectPtr holder = wrapTuple(index);
+
+  variable_list variableIndices;
```
Comment (killeent) [inline torch/csrc/autograd/python_variable_indexing.cpp]: I'm making the assumption that this section constitutes advanced indexing, but it would be helpful to have a code comment indicating that this is where those semantics begin.

**Review by killeent — COMMENTED** [review]
Review Part I - I want to split this into parts so I can better understand certain things before continuing on with my review.

File: aten/src/ATen/native/Indexing.cpp [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+#include <algorithm>
+#include <functional>
+#include <numeric>
+#include <vector>
+
+namespace at { namespace native {
+
+static bool isByteTensor(const Tensor & t) {
+  return t.defined() && t.type().scalarType() == kByte;
```
[the remark this answers is not in the document]
Comment (killeent) [reply aten/src/ATen/native/Indexing.cpp]: Oops. I mean the type check, i.e. if the line was just:

```
return t.type().scalarType() == kByte;
```

[the remark this answers is not in the document]
Comment (colesbury) [reply aten/src/ATen/native/Indexing.cpp]: Oh, good point. I think that will work now with Greg's `UndefinedType` change.

File: torch/csrc/autograd/python_variable_indexing.cpp (hunk 2 above)
Comment (killeent) [reply torch/csrc/autograd/python_variable_indexing.cpp]: Okay, can we add that as a comment. In general having negative return values for these errors seems like it happens in Python, but its a little mentally incongruous when the operation itself is getting what we would expect to be a non-negative number.

**Comment by fmassa** [comment]
This looks very nice!

Super clean implementation that extends the previous behavior, and I believe we have numpy parity on advanced indexing now!
I found particularly interesting that you fused the boolean indexing together with list indexing (extending thus boolean indexing to work as in numpy without much effort). Also, it seems that we get autograd (and unlimited backprop?) !

One question:
I have the impression that the old `masked_*` kernels are not used anymore. So for the case where we use a mask of the same size of the input (of dimension `d`), we perform a `nonzero` operation, `d` sums (for linearizing the indices) and then `take`/`put`, is that right?
Out of curiosity, do you have an idea on the performance impact of this change (specially for cuda tensors)? I personally think that having a performance impact here is worth the benefits, because of a super clean and easy to understand implementation with autograd support.

PS: It might be worth considering making `nonzero` return a tuple of 1d tensors, instead of a 2d tensor as discussed in https://github.com/pytorch/pytorch/issues/1834 . This behavior is used in this PR in https://github.com/pytorch/pytorch/pull/3725/files#diff-3e60b06052f270a7f3b76038f133a10cR65 , and might be convenient for users to have it like that. This is a breaking change though.

**Comment by colesbury** [comment]
There's some performance hit for not using `masked_select` 394us -> 638us:

https://gist.github.com/colesbury/de7193e8dd6ef92b4011a0c4c512096f

We could probably use `masked_select` for the one `ByteTensor` case. It's just a little tricky to make sure the output has the correct shape

File: torch/csrc/Exceptions.h [hunk 8]
```diff
@@ -14,6 +13,10 @@
 #define END_HANDLE_TH_ERRORS_RET(retval)                                       \
   } catch (python_error &e) {                                                  \
     return retval;                                                             \
+  } catch (torch::PyTorchError &e) {                                           \
```
Comment (apaszke) [inline torch/csrc/Exceptions.h]: a nicer solution would be to call it `set_python_error` or rename `python_error` to `error_already_set` (like `pybind` does)

File: torch/csrc/autograd/python_variable_indexing.cpp [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+    }
+  }
+  return count;
+}
+
+[[noreturn]]
+static void invalid_index(PyObject* obj) {
+  throw IndexError(
+    "only integers, slices (`:`), ellipsis (`...`), None and long or byte "
+    "Variables are valid indices (got %s)", Py_TYPE(obj)->tp_name);
```
Comment (apaszke) [inline torch/csrc/autograd/python_variable_indexing.cpp]: I think you should still mention that they have to be on a correct device

**Review by apaszke — COMMENTED** [review]
Haven't done a complete review, but I spotted a few things

File: aten/src/ATen/native/Indexing.cpp [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+
+  // Compute the linear index by multiplying the indexing tensors by the
+  // stride and summing them. All the indexing tensors have the same shape at
+  // this point. We also compute the number of dimensions before and after that
+  // are not being index.
+  Tensor linearIndex;
+  int64_t emptyBefore = 0, emptyAfter = 0, nElemBefore = 1, nElemAfter = 1;
+  for (int64_t i = 0; i < src.dim(); i++) {
+    if (indices[i].defined()) {
+      Tensor index = indices[i] * strides[i];
```
Comment (killeent) [inline aten/src/ATen/native/Indexing.cpp]: How does this operator overload work? From my interpretation, if `indices` is on the GPU this will invoke a `cudaMemcpy` and `mm` kernel launch for each dimension correct?

File: aten/src/ATen/native/Indexing.cpp [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+      nElemBefore *= src.size(i);
+    }
+  }
+
+  Type& longType = src.type().toScalarType(kLong);
+
+  // Compute the linear indices for the parts of the tensor not being indexed
+  Tensor beforeIndex;
+  if (emptyBefore > 0) {
+    auto index = longType.arange(0, nElemBefore) * strides[emptyBefore - 1];
```
Comment (killeent) [inline aten/src/ATen/native/Indexing.cpp]: Nice compression of the unindexed dimensions here!

**Review by killeent — COMMENTED** [review]
Pass II - I'll take a look at the set value logic and add overall thoughts in pass 3.

File: aten/src/ATen/native/Indexing.cpp (hunk 10 above)
Comment (colesbury) [reply aten/src/ATen/native/Indexing.cpp]: It's multiplying the indices (a LongTensor, maybe CUDA) by an integer. It translates to `mul` at the TH-level. So just a single kernel launch and no extra memcpy.

File: tools/jit/gen_jit_dispatch.py [hunk 12]
```diff
@@ -85,7 +85,9 @@ def is_tensor_arg(arg):
         # TensorList arguments, and functions that are only available as methods.
         if 'namespace' in decl['method_of']:
             if any(arg['simple_type'] == 'TensorList' for arg in arguments):
-                assert sum(map(is_tensor_arg, arguments)) == 1
+                if sum(map(is_tensor_arg, arguments)) != 1:
+                    # TODO: support this
```
[the remark this answers is not in the document]
Comment (colesbury) [reply tools/jit/gen_jit_dispatch.py]: Because indexing has a TensorList and one or two Tensor argument. Someone else (either you or Ed) needs to add JIT support for indexing.

File: torch/csrc/Exceptions.h (hunk 8 above)
Comment (colesbury) [reply torch/csrc/Exceptions.h]: Either is fine with me. Which do you prefer?

File: torch/csrc/autograd/python_variable_indexing.cpp (hunk 9 above)
Comment (colesbury) [reply torch/csrc/autograd/python_variable_indexing.cpp]: There's a different error message when things aren't on the correct devices. This one is jsut about types.

(We also don't mention other restrictions such as the indices have to be in-bounds and broadcastable)

File: torch/csrc/autograd/python_variable_indexing.cpp [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+    PyObject* obj = PyTuple_GET_ITEM(index, i);
+    if (THPUtils_checkLong(obj)) {
+      result = applySelect(result, dim, THPUtils_unpackLong(obj));
+    } else if (PySlice_Check(obj)) {
+      result = applySlice(result, dim, obj);
+      dim++;
+    } else if (obj == Py_Ellipsis) {
+      int64_t unspecified_dims = self.dim() - count_specified_dimensions(index);
+      for(; unspecified_dims > 0; unspecified_dims--) {
+        dim++;
```
[the remark this answers is not in the document]
Comment (colesbury) [reply torch/csrc/autograd/python_variable_indexing.cpp]: Whoops...

File: aten/src/ATen/native/Indexing.cpp (hunk 10 above)
Comment (fmassa) [reply aten/src/ATen/native/Indexing.cpp]: I believe `strides[i]` is a number, so it performs a `cmul`, but indeed, there is a kernel launch for every dimension.

File: torch/csrc/Exceptions.h (hunk 8 above)
Comment (apaszke) [reply torch/csrc/Exceptions.h]: I'm also fine with both of them. `error_already_set` seems a bit cleaner, but it's a larger change so I understand if you don't want to do it

File: tools/jit/gen_jit_dispatch.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (apaszke) [reply tools/jit/gen_jit_dispatch.py]: Can you please open an issue? It's easy to forget about a TODO in the code like this

**Review by killeent — COMMENTED** [review]
Okay, the `set` logic looks reasonable too. So given that this operates on variables, there is currently a mismatch in parity between what Tensor indexing allows and Variable indexing allows correct?

**Comment by colesbury** [comment]
@killeent, yeah there's going to be a mismatch until we merge Variable and Tensor

**Comment by killeent** [comment]
Okay cool. One other thing I would check is newaxis position correctness (see the test case here: https://github.com/pytorch/pytorch/pull/2779/files). If you are taking NumPy's implementation directly it should work I'm assuming but worth looking into.

**Comment by dzhulgakov** [comment]
@colesbury - looks like it completely broke symbolic export as now majority of indexing operations get traced as as_strided operation with hard-coded strides. Instead we need to capture the original indexing operations, so x[0] becomes something like Slice[start=0, end=0] instead of hard-coded dimensions of the tensor.

Any suggestions on what's a sustainable way of fixing it? At least the non-strided version of indexing needs to work.
